A Whole bunch of refactors.  See each commit

bf4 opened the pull request:
Can be divided conceptually into connection building and query building

re: connection building, these can be used for some api calls not in this gem, such as invoice#deliver invoice#pdf, etc. An example of this is in the attachment behavior

```ruby
def get_pdf(quickbooks_id)
  headers = { 'Content-Type' => 'application/pdf', 'Accept' => 'application/pdf' }
  connection = build_connection(@endpoint_url, headers: headers) do |conn|
    add_exception_middleware(conn)
    conn.request :oauth2, @access_token, token_type: 'bearer'
    add_connection_adapter(conn)
  end
  path = "#{entity_path(:invoice)}/#{quickbooks_id}/pdf"
  raw_response = raw_request(:get, conn: connection, path: path)
  raw_response.body
end

def deliver(quickbooks_id, send_to: nil, params: nil)
  headers = { 'Content-Type' => 'application/octet-stream' }
  connection = authorized_json_connection(@endpoint_url, headers: headers, middleware_builder: proc do |conn|
    add_exception_middleware(conn)
    add_connection_adapter(conn)
  end)
  params ||= {}
  params[:sendTo] = send_to if send_to
  path = "#{entity_path(:invoice)}/#{quickbooks_id}/send"
  raw_response = raw_request(:get, conn: connection, path: path, params: params)
  raw_response.body.fetch(singular(entity))
end
```

re query building, this makes it possible to extend the query building pieces like

```ruby
      # @example
      #   fields: "Id, FullyQualifiedName"
      #   conditions: { DocNumber: 2133 }             # aka filterable
      #   raw_condition: "DisplayName LIKE 'ACME%'"
      #   order_by: { DisplayName: :desc }            # aka sortable
      #   page_size: 1
      #   num_pages: 1
      #   inactive: true
      #   params: { minorversion: 12 }
      def where(entity, fields: nil, conditions: {}, raw_condition: "", order_by: default_order_by(entity), page_size: 1000, num_pages: :all, inactive: false, params: nil)
        entity or fail ArgumentError, "Entity cannot be nil"
        params ||= {}
        params[:requestid] ||= uuid
        params[:minorversion] ||= minor_version
        max_records = num_pages == :all ? false : num_pages * page_size
        records = []
        query_string = query_string(entity, fields: fields, conditions: conditions, raw_condition: raw_condition, order_by: order_by)
        all(entity, max: page_size, select: query_string, inactive: inactive).each {|record|
          yield record if block_given?
          records << record
          break if max_records && records.size >= max_records
        }
        records
      end

      def offset_query(entity, limit:, offset:, fields: nil, conditions: {}, raw_condition: "", order_by: default_order_by(entity), params: nil)
        entity or fail ArgumentError, "Entity cannot be nil"
        params ||= {}
        params[:requestid] ||= uuid
        params[:minorversion] ||= minor_version
        offset = offset + 1 # qb offsets start at one, but our convention is 0
        query_string = query_string(entity, fields: fields, conditions: conditions, raw_condition: raw_condition, order_by: order_by)
        query_string = offset_query_string(query_string, limit: limit, offset: offset)
        Array(query(query_string, params: params)).each do |record|
          yield record if block_given?
        end
      end

      # default order by for sane pagination
      def default_order_by(_entity)
        { "MetaData.CreateTime" => :desc }
      end

      def query_string(entity, fields: nil, conditions: {}, raw_condition: "", order_by: {})
        select_clause = build_all_query(entity, select: "SELECT #{fields || '*'} FROM #{singular(entity)}")
        conditions = conditions.map {|key, value|
          "#{key} = '#{esc(String value)}'"
        }.join(" AND ")
        where_clause =
          if conditions.empty?
            nil
          else
            conditions = "WHERE #{conditions}"
            build_all_query(entity, select: conditions)
          end
        raw_clause =
          if raw_condition.empty?
            nil
          else
            where_clause.nil? ? "WHERE #{raw_condition}" : "AND #{raw_condition}"
          end
        raw_order_by =
          if order_by.empty?
            nil
          else
            "ORDERBY" << order_by.map {|field_name, direction|
              " #{field_name} #{direction.to_s.upcase}"
            }.join(",")
          end
        [select_clause,where_clause,raw_clause,raw_order_by].compact.join(" ")
      end

```

minimul commented:
Thanks @bf4 

@ntippie Would you help review these refactors and pr #66  Thanks in advance

bf4 commented on lib/qbo_api.rb:
```diff
@@ -104,40 +114,65 @@ def reconnect
+    entity ? entity_response(data, entity) : data
   rescue => e
-    # Catch fetch key errors and just return JSON
+    # NOTE(BF): Rescue all StandardError and return parsed JSON body
```
best view this in split view.  I don't like this timid code which rescues all StandardError.  Related, to JSON requests perhaps we can add the `conn.use FaradayMiddleware::ParseJson, parser_options: { symbolize_names: false }` middleware.

Relatedly, I'd like to wrap json and xml encoding/decoding rather than specifying the library

bf4 commented on lib/qbo_api.rb:
```diff
@@ -104,40 +114,65 @@ def reconnect
+    case resp.headers['Content-Type']
+    when /json/ then JSON.parse(body)
+    else body
+    end
```
could rescue `JSON::ParserError` if we want.  notice I switch on the response content type header.  This might not always work since quickbooks api has some quirks where they don't follow http standards

bf4 commented on lib/qbo_api.rb:
```diff
@@ -104,40 +114,65 @@ def reconnect
+      # e.g. {"QueryResponse":{"totalCount":29}, "time":"2015 -11-05T04:30:59.195-08:00"}
+      entity_body = data['QueryResponse']
+      # NOTE(BF): convert an empty QueryResponse into nil
+      return nil if entity_body.empty?
```
not sure why we'd want to convert this to nil

bf4 commented on lib/qbo_api.rb:
```diff
@@ -104,40 +114,65 @@ def reconnect
-      ar.first.fetch(singular(entity))
+    if data.key?('QueryResponse')
+      # e.g. {"QueryResponse":{"totalCount":29}, "time":"2015 -11-05T04:30:59.195-08:00"}
+      entity_body = data['QueryResponse']
```
not sure why we'd return the whole data when the entity_body doesn't have an entity_name in it

bf4 commented on lib/qbo_api.rb:
```diff
@@ -104,40 +114,65 @@ def reconnect
+    entity_name = entity_name(entity)
+    entity_body.fetch(entity_name) do
+      QboApi.logger.debug { "[Quickbooks] entity name not in response body: entity=#{entity.inspect} entity_name=#{entity_name.inspect} body=#{data.inspect}" }
+      # NOTE(BF): Return full response body when entity name not in entity body
```
not sure why we'd return the whole data when the entity_body doesn't have an entity_name in it

bf4 commented:
rebased off of master and cleaned up some more. I'm not a big fan of modules as code organization (i.e. they depend on what they're included in), but I'm following that pattern here, and I think it could help

bf4 commented:
Updated post #66

bf4 commented:
@minimul would you like me to pull out a few these commits into their own PRs? e.g. e8a3e62

bf4 commented:
did it anyways

ntippie commented:
Added comments to the individual PRs. Thank you, @bf4!

bf4 commented:
Here's what's left

minimul commented on lib/qbo_api/api_methods.rb:
```diff
@@ -19,24 +19,30 @@ def query(query, params: nil)
       end
     end
 
+    # @example
```
Remove all comments.

minimul commented on lib/qbo_api/api_methods.rb:
```diff
@@ -19,24 +19,30 @@ def query(query, params: nil)
 
     # @example
```
Remove all comments.

minimul commented:
@ntippie Can you review the connection building commit. Thanks.

minimul commented:
This and the other PRs officially released as v1.8.0

@bf4 You have greatly improved this library.  Many thanks.
If your time allows it I look forward to more enchantments and improvements.

@ntippie Many thanks for your fast and insightful input.

bf4 commented:
@minimul @ntippie Been fun working with you.  I do have more to add, and I'll be adding tests for those :) like `get_invoice_pdf` and `deliver_invoice` and separating out the different auth concerns.

bf4 commented:
e.g. see #76
